README.md
A library for Dart developers that makes [SAP BusinessOne](https://www.sap.com/products/business-one.html)(B1) Service Layer (SL) programming a lot simpler than using plain http package. B1 is one of the most popular information systems for small and mid sized companies. It runs on Microsoft SQL Server as well as on SAP HANA. SL is available only for the HANA version. B1 is a commercial product, and if you want to be a developer, you should contact a SAP Partner or SAP to become a partner. This Dart package is only for developers who are part of this community. If you work for a company that already has B1, this package could be very helpful, too, for building custom applications, since then you are entitled to use SL.
[Official Service Layer documentation](https://help.sap.com/doc/0d2533ad95ba4ad7a702e83570a21c32/9.3/en-US/Working_with_SAP_Business_One_Service_Layer.pdf) is available only if you have an S user ID. On the other hand, a six part [Service Layer API video](https://www.youtube.com/watch?v=zaF_i7x9-s0&list=PLMdHXbewhZ2QsgYSICRQuoL8lkoEHjNzS&index=22) is publicly available.

The f in the name was inspired by the name sqflite.

Created from templates made available by Stagehand under a BSD-style
[license](https://github.com/dart-lang/stagehand/blob/master/LICENSE).

SAP, SAP Business One are registered trademarks of [SAP](https://www.sap.com/corporate/en/company.html)  

## Usage

For a full working example review the example folder as well as the nice collection of automated Dart test scripts in the test folder.
Here is a typical skeleton of an application using the functionality of the package:

```dart
import 'dart:convert';
import 'package:b1serviceflayer/b1serviceflayer.dart';

main() async {
  ...
  final b1s = B1ServiceLayer(B1Connection(serverUrl: url, 
    companyDB: companyDB, userName: user, password: pwd));
  try {    
    await b1s.loginAsync().timeout(const Duration(seconds: 10));
    ...
    String activitiesJson = await b1s.queryAsync("Activities");
    Map<String, dynamic> activitiesMap = json.decode(activitiesJson);
    ...
    await b1s.updateAsync(entityName: "Activities($activityCodeToComplete)",
      entityJSON: json.encode(activityList[0]));
    ...
    String activityJson = await b1s.queryAsync("Activities($activityCodeToComplete)");
    ...
    activityJson =   return await b1s.createAsync(entityName: "Activities",
      entityJSON: json.encode(activityData));
    ...
    await b1s.deleteAsync(entityName: "Activities($activityCodeToComplete)",
      errorWhenDoesntExist: true);
    ...
    await b1s.logoutAsync();
  } on B1Error catch(exception) {
      print("Exception is B1Error (${exception.statusCode}) ${exception.error.message.value} (${exception.error.code}) for Query ${exception.queryUrl}");
      print("Payload ${exception.postBody}");
  } catch(exception,stackTrace) {
    print(exception); print(stackTrace);
  }
}
```
All SL invocation is performed via the regular, and brilliant, Dart [http](https://pub.dev/packages/http) package.
All functions return a Future, which can be used with the regular await keywords in an async body; this is the reason all function names is suffixed with Async as reminder, that these should be invoket with await, or the classic Future handling ways.

The main class is **B1ServiceLayer**, which requires a **B1Connection** object. The B1Connection is initialiyed with the four mandatory SL login parameters.
Here is the list of the functions:
- **loginAsync** the library automatically calls login when the session is expired after 30 minutes. So, all subsequent calls of the same service layer object use the same connection, but after 30 minutes, a relogin is automatically performed. 
- **queryAsync** returns a list of objects or a single one, depending on the query string.
- **createAsync** creates an B1 entity object in B1, and returns the full details of the created object.
- **updateAsync** updates a B1 entity object. Make sure to pass only the values that you really want to update, otherwise you may overwrite some valuable data by mistake.
- **deleteAsync** deletes a B1 entity object. Very few objects are allowed to be deleted in SAP. Activities are very good for tests, since they are deletable.
- **logoutAsync** Sessions are automatically timeouted in B1 SL, so calling logout is required only when the user is changed.

Only *queryAsync* and *createAsync* have return value, a JSON string, which can be parsed with the dart **json.decode** function into a **Map<String,dynamic>**. 
The functions *createAsync* and *updateAsync* are expecting JSON strings as input parameters. The **json.encode** function can be used to concvert Map<String,dynamic> objects into JSON strings.
These functions are meant for rudimentary JSON communication, they don't parse the JSON string, except the error responses. This layer gives the ultimate performance, as much as possible with the SL and Dart infrastructure. If you want a higher level of convenience you can use JSON Dart conversion generators. For example, the B1Error class was generated with [quicktype](https://app.quicktype.io), but other tools can be used either. Pay attention, however that the B1 JSON structures are big, and a performance may be a problem with these code generators.  

In future versions additional functions are added for batch execution, image uploading.

You can use the Future's **timeout** to control the tolerable duration in your application. I was experimenting with using the *connectionTimeout* of the HttpClient of dart:io, but the Future timeout is a lot more convenient.
When SL returns an error response, B1ServiceLayer functions throw **B1Error**. Upon timeout or network errors the corresponding exceptions are thrown by the underlying Dart machinery. B1Error has a nice bunch of properties for error analysis, which makes programming clean.

## Performance
Here is a log of the sample application from the example folder:
```
Logging in ...
Querying activities ...
9 activities returned in 5274 ms
Activity 8 Completed -3 Notes Completed on 2019-08-05 21:17:13.951161 in 53 ms
Activity created with code 15 Notes A new job to do in 147 ms
Querying activities with newly added activity...
Number of activities 9 returned in 102 ms
```
The first operation was slow, even after an explicit login, but then the performance was quite acceptable.

## License
The license is an open source BSD license as suggested by the [Dart package publishing documentation](https://dart.dev/tools/pub/publishing), since Dart itself has BSD license. 

See the [LICENSE.md](https://github.com/nemethmik/b1serviceflayer/blob/master/LICENSE.md) file.

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/nemethmik/b1serviceflayer


CHANGELOG.md
## 1.0.0

- The project was initated with Stagehand, and then a nicely working version is available with this version. Tests have been added as well as an example. Breaking changes are not expected. 

## 1.0.1

- Documentation has been extended significantly.
- Since logins are so important to monitor, a **printLogins** optional bool parameter added to the B1ServiceLayer constructor, which enables printing a line in the console for debugging purposes.
- **executionMilliseconds** property added to B1ServiceLayer to help monitoring performance, which is so crytical in mobile applications.
- The example and some of the test scripts have been adjusted accordingly.

## 1.0.2

- tYES, tNO relocated into a new BoYesNoEnum class.
- BoEnums abstract class was added as a base class for defining typesafe string enumeration types.

## 1.0.3

- Unfortunately, flutter_test depends on meta 1.1.6 today 2019.08.09 and I decremented the meta library version to 1.1.6.

test/b1login_test.dart
import 'dart:async';
import 'dart:io';
import 'dart:convert';
import 'package:test/test.dart';
import "package:http/http.dart" as http;
import 'package:b1serviceflayer/b1serviceflayer.dart';
import 'logindetailsfortests.dart' as conf;

void main() {
  group("B1SL_LoginTestSeries", () {
    test("SuccessfulLogin",() async {
      // final b1s = B1Services(serverUrl: ipAddress + url, companyDB: companyDB, userName: user, password: pwd);
      final b1s = B1ServiceLayer(B1Connection(serverUrl: conf.ipAddress + conf.url,
        companyDB: conf.companyDB, userName: conf.user, password: conf.pwd));
      try {
        final sessionId = await b1s.loginAsync().timeout(const Duration(seconds: 15));
        print("Execution time ${b1s.exetutionMilliseconds}");
        expect(sessionId.b1connection,isNotNull);
      } on B1Error catch(exception) {
        print("Exception is B1Error (${exception.statusCode}) ${exception.error.message.value} (${exception.error.code}) for Query ${exception.queryUrl}");
        print("Payload ${exception.postBody}");
        expect(exception,isNull);
      } on http.Response catch(exception) {
        print("Exception is HTTP.Response ${exception.body}");
        expect(exception,isNull);
      } catch(exception,stackTrace) {
        print(exception);
        print(stackTrace);
        expect(exception,isNull);
      }
    }); 
    test("LoginAttemptWithInvalidUser",() async {
      // final b1s = B1Services(serverUrl: ipAddress + url, companyDB: companyDB, userName: "krikszkraksz", password: pwd);
      final b1s = B1ServiceLayer(B1Connection(serverUrl: conf.ipAddress + conf.url,
        companyDB: conf.companyDB, userName: "krikszkraksz", password: conf.pwd));
      try {
        final sessionId = await b1s.loginAsync();
        expect(sessionId,isNull);
      } catch(exception) {
        if(exception is B1Error) {
          print("Exception is B1Error (${exception.statusCode}) ${exception.error.message.value} (${exception.error.code}) for Query ${exception.queryUrl} Payload ${exception.postBody}");
        }
        expect(exception is B1Error,isTrue);
      }
    }); 
    test("LoginAttemptWithInvalidURL", () async {
      //(401) Invalid session. (301)
      final b1s = B1ServiceLayer(B1Connection(serverUrl: conf.ipAddress + conf.url + "krikszkraksz",
        companyDB: conf.companyDB, userName: conf.user, password: conf.pwd));
      try {
        await b1s.loginAsync();
        expect(false,isTrue);
      } catch(exception) {
        if(exception is B1Error) {
          print("Exception is B1Error (${exception.statusCode}) ${exception.error.message.value} (${exception.error.code}) for Query ${exception.queryUrl} Payload ${exception.postBody}");
        }
        expect(exception is B1Error,isTrue);
      }
    });
    test("LoginAttemptWithInvalidCompanyDB", () async {
      final b1s = B1ServiceLayer(B1Connection(serverUrl: conf.ipAddress + conf.url,
        companyDB: "krikszkraksz", userName: conf.user, password: conf.pwd));
      try {
        await b1s.loginAsync();
        expect(false,isTrue);
      } catch(exception) {
        if(exception is B1Error) {
          print("Exception is B1Error (${exception.statusCode}) ${exception.error.message.value} (${exception.error.code}) for Query ${exception.queryUrl} Payload ${exception.postBody}");
        }
        expect(exception is B1Error,isTrue);
      }
    });
    test("LoginAttemptWithInvalidPassword", () async {
      final b1s = B1ServiceLayer(B1Connection(serverUrl: conf.ipAddress + conf.url,
        companyDB: conf.companyDB, userName: conf.user, password: "krikszkraksz"));
      try {
        await b1s.loginAsync();
      } catch(exception) {
        if(exception is B1Error) {
          print("Exception is B1Error (${exception.statusCode}) ${exception.error.message.value} (${exception.error.code}) for Query ${exception.queryUrl} Payload ${exception.postBody}");
        }
        expect(exception is B1Error,isTrue);
      }
    });
    test("LoginAttemptWithFaultyIPAddress", () async {
      //This is a long timeout error because of the underlying network layer's timeout
      final b1s = B1ServiceLayer(B1Connection(serverUrl: "http://192.168.250.92:50001" + conf.url,
        companyDB: conf.companyDB, userName: conf.user, password: conf.pwd));
      try {
        await b1s.loginAsync().timeout(const Duration(seconds: 5));
      } catch(error) {
        print(error);
        expect(error,isNotNull);
      }
    },timeout: Timeout(Duration(minutes: 5)));
    test("Login with HTTPClient",() async {
      //This is just a fun experiment with using the low level HttpClient
      try {
        final String postBody = '{"UserName":"${conf.user}", "Password":"${conf.pwd}", "CompanyDB":"${conf.companyDB}"}';
        final String queryUrl = conf.ipAddress + conf.url + "Login";
        //final String queryUrl = "http://192.168.250.92:50001" + url + "Login";
        final HttpClient client = HttpClient()..connectionTimeout = Duration(seconds: 15);
        final HttpClientRequest request = await client.postUrl(Uri.parse(queryUrl));
        request.write(postBody);
        final HttpClientResponse response = await request.close();
        dynamic responseBodyText = await _getResponseBody(response);
        print(responseBodyText);
        expect(responseBodyText, isNotNull);
        expect(responseBodyText.isNotEmpty, isTrue);
      } catch(error,stackTrace) {
        print(error);
        print(stackTrace);
        expect(error,isNull);
      }
    });
  });
}
Future<dynamic> _getResponseBody(HttpClientResponse response) {
  StringBuffer responseBody = StringBuffer();
  Completer completer = Completer();
  response.transform(utf8.decoder).listen((String onData){
    responseBody.write(onData);
  }, onDone: ()=> completer.complete(responseBody.toString()));
  return completer.future;
}


test/b1query_test.dart
import 'package:test/test.dart';
import 'package:b1serviceflayer/b1serviceflayer.dart';
import 'logindetailsfortests.dart' as conf;

void main() {
  group("Low Level SL Tests",() {
    final b1s = B1ServiceLayer(B1Connection(serverUrl: conf.ipAddress + conf.url,
      companyDB: conf.companyDB, userName: conf.user, password: conf.pwd));
    test("ErrorHandlingForInvalidPropertyName", () async {
      try {
        await b1s.queryAsync("Items?\$select=ItemCodeX,ItemName");
      } catch(e) {
        if(e is B1Error) {
          print("Exception is B1Error (${e.statusCode}) ${e.error.message.value} (${e.error.code}) for Query ${e.queryUrl} Payload ${e.postBody}");
        }
        const expectedErrorCode = -1000; //Property ItemCodeX of Item is invalid
        expect(b1s.b1Error.error.code, equals(expectedErrorCode));
      }
    });
    test("ErrorHandlingForInvalidDocStatusCode", () async {
      try {
        await b1s.queryAsync("PurchaseOrders?\$filter=DocumentStatus eq 'bost_OpenX'");
      } catch(e) {
        if(e is B1Error) {
          print("Exception is B1Error (${e.statusCode}) ${e.error.message.value} (${e.error.code}) for Query ${e.queryUrl} Payload ${e.postBody}");
        }
        const expectedErrorCode = -1013; // Invalid item name bost_OpenX in Enum BoStatus The valid names are: bost_Open-O, bost_Close-C, bost_Paid-P, bost_Delivered-D
        expect(b1s.b1Error.error.code, equals(expectedErrorCode));
      }
    });
  });
}

test/b1activity_test.dart
import 'dart:convert';
import 'package:test/test.dart';
import 'package:b1serviceflayer/b1serviceflayer.dart';
import 'logindetailsfortests.dart' as conf;
void main() {
  group("Low Level SL Activity Queries",(){
    final b1s = B1ServiceLayer(B1Connection(serverUrl: conf.ipAddress + conf.url,
      companyDB: conf.companyDB, userName: conf.user, password: conf.pwd));
    test("QueryAllActivities", () async {
      try {
        final queryResponseJSON = await b1s.queryAsync("Activities?\$select=ActivityCode,HandledByEmployee,Details");
        Map<String, dynamic> queryResponseMap = json.decode(queryResponseJSON);
        print("Execution millis ${b1s.exetutionMilliseconds}");
        expect(queryResponseMap.isNotEmpty,isTrue);
        expect(queryResponseMap["value"][0]["ActivityCode"], greaterThan(0));
      } catch(e,stackTrace) {
        if(e is B1Error) {
          print("Exception is B1Error (${e.statusCode}) ${e.error.message.value} (${e.error.code}) for Query ${e.queryUrl} Payload ${e.postBody}");
        } else {
          print(e); print(stackTrace);
        }
        expect(e,isNull);
      }
    });
  });
}

lib/src/quicktype_error.dart
import 'dart:convert';
import 'package:meta/meta.dart';
class B1Error {
  int statusCode;//Manually added for convenience, not coming from SAP B1
  String queryUrl;
  String postBody;
  void setStatusCode(int statusCode,{@required String queryUrl,String postBody}) {
    this.statusCode = statusCode;
    this.queryUrl = queryUrl;
    this.postBody = postBody;
  }
  Error error;
  B1Error({
      this.error,
  });
  factory B1Error.fromJson(String str) => B1Error.fromMap(json.decode(str));
  String toJson() => json.encode(toMap());
  factory B1Error.fromMap(Map<String, dynamic> json) => B1Error(
      error: json["error"] == null ? null : Error.fromMap(json["error"]),
  );
  Map<String, dynamic> toMap() => {
      "error": error == null ? null : error.toMap(),
  };
}

class Error {
    int code;
    Message message;
    Error({
        this.code,
        this.message,
    });
    factory Error.fromJson(String str) => Error.fromMap(json.decode(str));
    String toJson() => json.encode(toMap());
    factory Error.fromMap(Map<String, dynamic> json) => Error(
        code: json["code"] == null ? null : json["code"],
        message: json["message"] == null ? null : Message.fromMap(json["message"]),
    );
    Map<String, dynamic> toMap() => {
        "code": code == null ? null : code,
        "message": message == null ? null : message.toMap(),
    };
}

class Message {
    String lang;
    String value;
    Message({
        this.lang,
        this.value,
    });
    factory Message.fromJson(String str) => Message.fromMap(json.decode(str));
    String toJson() => json.encode(toMap());
    factory Message.fromMap(Map<String, dynamic> json) => Message(
        lang: json["lang"] == null ? null : json["lang"],
        value: json["value"] == null ? null : json["value"],
    );
    Map<String, dynamic> toMap() => {
        "lang": lang == null ? null : lang,
        "value": value == null ? null : value,
    };
}


lib/src/quicktype_session.dart
import 'dart:convert';
//Generated with https://app.quicktype.io/ from assets/login_response.json
class Session {
    final String odataMetadata;
    final String sessionId;
    final String version;
    final int sessionTimeout;
    Session({
        this.odataMetadata,
        this.sessionId,
        this.version,
        this.sessionTimeout,
    });
    factory Session.fromJson(String str) => Session.fromMap(json.decode(str));
    String toJson() => json.encode(toMap());
    factory Session.fromMap(Map<String, dynamic> json) => Session(
        odataMetadata: json["odata.metadata"] == null ? null : json["odata.metadata"],
        sessionId: json["SessionId"] == null ? null : json["SessionId"],
        version: json["Version"] == null ? null : json["Version"],
        sessionTimeout: json["SessionTimeout"] == null ? null : json["SessionTimeout"],
    );
    Map<String, dynamic> toMap() => {
        "odata.metadata": odataMetadata == null ? null : odataMetadata,
        "SessionId": sessionId == null ? null : sessionId,
        "Version": version == null ? null : version,
        "SessionTimeout": sessionTimeout == null ? null : sessionTimeout,
    };
}


lib/src/b1serviceflayer_base.dart
import 'dart:async';
import 'dart:io';
import "package:meta/meta.dart";
import "package:http/http.dart" as http;
import "./quicktype_session.dart";
import "./quicktype_error.dart";
class B1Connection {
  final String serverUrl;
  final String userName;
  final String password;
  final String companyDB;
  Session get b1session => _b1session; Session _b1session;
  String get b1Cookies => _b1Cookies; String _b1Cookies;
  DateTime get loginTime => _loginTime; DateTime _loginTime;
  B1Connection({@required this.serverUrl, @required this.userName, @required this.password, @required this.companyDB});
  bool get isSessionExpired => loginTime == null || b1session == null || (DateTime.now().millisecondsSinceEpoch - loginTime.millisecondsSinceEpoch) >= (b1session.sessionTimeout * 60 * 1000);
  _setSession(Session session,String cookies) {
    _b1session = session;
    //b1session.sessionTimeout ??= 30 * 60 * 1000; // Same as if b1session.sessionTimeout == null then session.sessionTimeout = value
    _loginTime = DateTime.now();
    _b1Cookies = cookies;
  }
}
class B1ServiceLayer {
  final B1Connection b1connection;
  String get queryUrl => _queryUrl; String _queryUrl;
  http.Response get queryResponse => _queryResponse; http.Response _queryResponse;
  B1Error get b1Error => _b1Error; B1Error _b1Error;
  bool get hasError => b1Error != null;
  int get exetutionMilliseconds => _exetutionMilliseconds; int _exetutionMilliseconds;
  final bool printLogins; 
  B1ServiceLayer(this.b1connection,{this.printLogins = false});
  Future<B1ServiceLayer> loginAsync() async {
    _b1Error = null;
    _queryUrl = b1connection.serverUrl + "Login";
    String postBody = '{"UserName":"${b1connection.userName}", "Password":"${b1connection.password}", "CompanyDB":"${b1connection.companyDB}"}';
    int start = DateTime.now().millisecondsSinceEpoch;
    _queryResponse = await http.post(queryUrl,body:postBody);
    _exetutionMilliseconds = DateTime.now().millisecondsSinceEpoch - start;
    if(queryResponse.statusCode == HttpStatus.ok) {
      b1connection._setSession(Session.fromJson(queryResponse.body),queryResponse.headers["set-cookie"]); 
      if(printLogins) {
        print("B1ServiceLayer.Login at ${b1connection.loginTime.millisecondsSinceEpoch} in $exetutionMilliseconds ms");
      }
      //loginResponse.headers["set-cookie"] is empty in a browser (client)
      //b1session.setLoggedIn(loginResponse.headers["set-cookie"], loginResponse.data.SessionTimeout);
      //print("B1SLServices.loginAsync:" + queryResponse.body);
    } else {
      if(queryResponse.body != null) {
        try { 
          _b1Error = B1Error.fromJson(queryResponse.body) 
            ..statusCode = queryResponse.statusCode
            ..postBody = postBody
            ..queryUrl = queryUrl;
          throw b1Error;
        } finally{}
      }
      throw queryResponse;
    }
    return this;
  }
  Future<bool> logoutAsync() async {
    if(b1connection.isSessionExpired) { 
      return true;
    } else {
      _queryUrl = b1connection.serverUrl + "Logout";
      _b1Error = null;  
      int start = DateTime.now().millisecondsSinceEpoch;
      _queryResponse = await http.get(queryUrl, headers: {HttpHeaders.cookieHeader: b1connection.b1Cookies});
      _exetutionMilliseconds = DateTime.now().millisecondsSinceEpoch - start;
      if(queryResponse.statusCode == HttpStatus.noContent) {
        b1connection._loginTime = null; //Mark logged out
        return true;
      } else {
        if(queryResponse.body != null) {
          try{
            _b1Error = B1Error.fromJson(queryResponse.body) 
              ..statusCode = queryResponse.statusCode
              ..queryUrl = queryUrl;
            throw b1Error;
          }finally{}
        }
        throw queryResponse;
      }
    }
  }
  Future<String> queryAsync(String queryString,{bool errorWhenNotFound = false}) async {
    if(b1connection.isSessionExpired) { await this.loginAsync();}
    _queryUrl = b1connection.serverUrl + queryString;
    _b1Error = null;  
    int start = DateTime.now().millisecondsSinceEpoch;
    _queryResponse = await http.get(queryUrl, headers: {HttpHeaders.cookieHeader: b1connection.b1Cookies});
    _exetutionMilliseconds = DateTime.now().millisecondsSinceEpoch - start;
    if(queryResponse.statusCode == HttpStatus.ok) {
      return queryResponse.body;
    } else {
      if(queryResponse.body != null) {
        if(queryResponse.statusCode == HttpStatus.notFound && !errorWhenNotFound) return null; 
        _b1Error = B1Error.fromJson(queryResponse.body) 
          ..statusCode = queryResponse.statusCode
          ..queryUrl = queryUrl;
        throw b1Error;
      }      
      throw queryResponse;
    }
  }
  Future<String> createAsync({@required String entityName, String entityJSON}) async {
    if(b1connection.isSessionExpired) {await loginAsync();}
    _queryUrl = b1connection.serverUrl + entityName;
    _b1Error = null;
    int start = DateTime.now().millisecondsSinceEpoch;
    _queryResponse = await http.post(queryUrl,body: entityJSON, headers: {HttpHeaders.cookieHeader: b1connection.b1Cookies});
    _exetutionMilliseconds = DateTime.now().millisecondsSinceEpoch - start;
    if(queryResponse.statusCode == HttpStatus.created) {
      return queryResponse.body;
    } else {
      if(queryResponse.body != null) {
        try {
          _b1Error = B1Error.fromJson(queryResponse.body) 
            ..statusCode = queryResponse.statusCode
            ..queryUrl = queryUrl
            ..postBody = entityJSON;
          throw b1Error;
        } finally {}
      }
      throw queryResponse;
    }
  }
  Future<void> updateAsync({@required String entityName, String entityJSON}) async  {
    if(b1connection.isSessionExpired){ await loginAsync();}
    _queryUrl = b1connection.serverUrl + entityName;
    _b1Error = null;  
    int start = DateTime.now().millisecondsSinceEpoch;
    _queryResponse = await http.patch(queryUrl,body:entityJSON, headers: {HttpHeaders.cookieHeader: b1connection.b1Cookies});
    _exetutionMilliseconds = DateTime.now().millisecondsSinceEpoch - start;
    if(queryResponse.statusCode == HttpStatus.noContent) {
    } else {
      if(queryResponse.body != null) {
        _b1Error = B1Error.fromJson(queryResponse.body)
          ..setStatusCode(queryResponse.statusCode,queryUrl: queryUrl,postBody: entityJSON);
        throw b1Error;
      }
      throw queryResponse;
    }
  }
  Future<void> deleteAsync({@required String entityName,bool errorWhenDoesntExist = false}) async {
    if(this.b1connection.isSessionExpired) {await loginAsync();}
    _queryUrl = b1connection.serverUrl + entityName;
    int start = DateTime.now().millisecondsSinceEpoch;
    _queryResponse = await http.delete(queryUrl,headers: {HttpHeaders.cookieHeader: b1connection.b1Cookies});
    _exetutionMilliseconds = DateTime.now().millisecondsSinceEpoch - start;
    _b1Error = null;  
    if(queryResponse.statusCode == HttpStatus.noContent) {
    } else {
      if(queryResponse.body != null) {
        _b1Error = B1Error.fromJson(queryResponse.body)
          ..setStatusCode(queryResponse.statusCode,queryUrl: queryUrl);
        var throwError = true;
        if(b1Error.error.code == -2028 && !errorWhenDoesntExist) {
          throwError = false;
        }
        if(throwError) {throw b1Error;}
        else {return null;}
      }
      throw queryResponse;
    }
  }  
}


example/b1serviceflayer_example.dart
import 'dart:convert';
import 'package:b1serviceflayer/b1serviceflayer.dart';

main() async {
  const url = "http://hana93srv:50001/b1s/v1/";
  const user = "manager";
  const pwd ="123qwe";
  const companyDB ="SBODEMOUS";
  final b1s = B1ServiceLayer(B1Connection(serverUrl: url, 
    companyDB: companyDB, userName: user, password: pwd),printLogins: true);
  try {
    print("Logging in ...");
    await b1s.loginAsync().timeout(const Duration(seconds: 10));

    print("Querying activities ...");
    String activitiesJson = await b1s.queryAsync("Activities");
    Map<String, dynamic> activitiesMap = json.decode(activitiesJson);
    List<dynamic> activityList = activitiesMap["value"];
    print('${activityList.length} activities returned in ${b1s.exetutionMilliseconds} ms');
    activityList.forEach((activity){
      print('Activity ${activity["ActivityCode"]} Completed ${activity["Status"]} Notes ${activity["Notes"]} ');
    });

    print("Updating activity ...");
    int activityCodeToComplete = activityList[0]["ActivityCode"];
    activityList[0]["Status"] = -3;
    activityList[0]["Notes"] = "Completed on ${DateTime.now()}";
    await b1s.updateAsync(entityName: "Activities($activityCodeToComplete)",
      entityJSON: json.encode(activityList[0]));

    print("Fetching updated activity ...");
    String activityJson = await b1s.queryAsync("Activities($activityCodeToComplete)");
    Map<String, dynamic> activityMap = json.decode(activityJson);
    print('Activity ${activityMap["ActivityCode"]} Completed ${activityMap["Status"]} Notes ${activityMap["Notes"]} in ${b1s.exetutionMilliseconds} ms');

    print("Adding a new activity ...");
    activityJson = await addActivity(b1s, 12);
    activityMap = json.decode(activityJson);
    print('Activity created with code ${activityMap["ActivityCode"]} Notes ${activityMap["Notes"]} in ${b1s.exetutionMilliseconds} ms');

    print("Deleting the new activity ...");
    activityCodeToComplete = activityMap["ActivityCode"];
    await b1s.deleteAsync(entityName: "Activities($activityCodeToComplete)", errorWhenDoesntExist: true);

    print("Querying activities with newly added activity...");
    activitiesJson = await b1s.queryAsync("Activities");
    activitiesMap = json.decode(activitiesJson);
    activityList = activitiesMap["value"];
    print('Number of activities ${activityList.length} returned in ${b1s.exetutionMilliseconds} ms');
    activityList.forEach((activity){
      print('Activity ${activity["ActivityCode"]} Completed ${activity["Status"]} Notes ${activity["Notes"]} ');
    });

    print("Logging out ...");
    await b1s.logoutAsync();
  } on B1Error catch(exception) {
      print("Exception is B1Error (${exception.statusCode}) ${exception.error.message.value} (${exception.error.code}) for Query ${exception.queryUrl}");
      print("Payload ${exception.postBody}");
  } catch(exception,stackTrace) {
    print(exception); print(stackTrace);
  }
}
Future<String> addActivity(B1ServiceLayer b1s, int userId) async {
  final Map<String,dynamic> activityData = {};
  //Subject:-1, ActivityType: -1 /*General*/, Status: -2 /*NotStarted, -3=Completed*/,
  activityData["Notes"] = "A new job to do";
  activityData["ActivityDate"] = "${DateTime.now()}";
  activityData["ActivityTime"] = "08:08:08"; //Activity creation date/time
  activityData["StartDate"] =  "${DateTime(2019,9,7)}"; //Automatically convert Dates and Times to strings in POST/PUT/PATCH
  activityData["StartTime"] = "10:10"; //UTC 0 offset, not the local time :(
  activityData["Details"] = "${DateTime.now()}: more explanation";
  activityData["ActivityType"] = -1;
  activityData["Activity"] = "cn_Task";
  activityData["Priority"] = "pr_High";
  activityData["PersonalFlag"] = BoYesNoEnum.tNO.value;
  activityData["DurationType"] = "du_Hours"; 
  activityData["Duration"] = 48.0;
  activityData["HandledBy"] =  userId; //A user number
  return await b1s.createAsync(entityName: "Activities",
    entityJSON: json.encode(activityData));
}